### PR TITLE
Add custom color parameters

### PR DESCRIPTION
**Pull request by Nutzzz**
/textcolor "RRGGBB" (default="000000")
/bgcolor "RRGGBB" (default="F4F4F4")

---
This change is [Reviewable](https://reviewable.io/reviews/digitalmediaserver/nsis-inetc-plugin/1)

### CODE REVIEW

**Comment by Nutzzz** [comment]
(For what it's worth)  We use UltraModernUI for NSIS with different skins for different flavors of our installer, so I needed an easy way to recolor based on the flavor, so I whipped a parameter-ized version of rmangroliya's snippet from here: http://forums.winamp.com/showthread.php?t=334819

Now I just add "/textcolor ${UMUI_TEXT_COLOR} /bgcolor ${MUI_BGCOLOR}" to my inetc call.

**Comment by Nadahar** [comment]
@Nutzzz This sounds like a nice addition. I'd just like to make sure that you've noticed that this isn't the "official" InetC plugin, but a fork I've made with an alternative download dialog.

If your modification is intended for the "original" it needs to be posted elsewhere.

I'll do a review using Reviewable.

**Comment by Nadahar** [comment]
@Nutzzz Even when building in ANSI mode I get errors with VS2012 (which is what has been used for this plugin: 
```PowerShell
1>inetc.obj : error LNK2001: unresolved external symbol _purecall
1>inetc.obj : error LNK2019: unresolved external symbol "void * __cdecl operator new(unsigned __int64)" (??2@YAPEAX_K@Z) referenced in function "char * __cdecl std::_Allocate<char>(unsigned __int64,char *)" (??$_Allocate@D@std@@YAPEAD_KPEAD@Z)
1>inetc.obj : error LNK2019: unresolved external symbol "void __cdecl operator delete(void *)" (??3@YAXPEAX@Z) referenced in function "public: __cdecl std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> >::~basic_string<char,struct std::char_traits<char>,class std::allocator<char> >(void)" (??1?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@QEAA@XZ)
1>inetc.obj : error LNK2019: unresolved external symbol atexit referenced in function "void __cdecl `dynamic initializer for 'public: static class std::_Generic_error_category std::_Error_objects<int>::_Generic_object''(void)" (??__E?_Generic_object@?$_Error_objects@H@std@@2V_Generic_error_category@2@A@@YAXXZ)
1>inetc.obj : error LNK2019: unresolved external symbol memcpy referenced in function "public: void __cdecl std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> >::_Copy(unsigned __int64,unsigned __int64)" (?_Copy@?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@QEAAX_K0@Z)
1>inetc.obj : error LNK2019: unresolved external symbol strlen referenced in function "public: __cdecl std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> >::basic_string<char,struct std::char_traits<char>,class std::allocator<char> >(char const *)" (??0?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@QEAA@PEBD@Z)
1>inetc.obj : error LNK2019: unresolved external symbol memmove referenced in function "public: class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > & __cdecl std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> >::erase(unsigned __int64,unsigned __int64)" (?erase@?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@QEAAAEAV12@_K0@Z)
1>inetc.obj : error LNK2019: unresolved external symbol sscanf_s referenced in function "__int64 __cdecl dlgProc(struct HWND__ *,unsigned int,unsigned __int64,__int64)" (?dlgProc@@YA_JPEAUHWND__@@I_K_J@Z)
1>inetc.obj : error LNK2019: unresolved external symbol "void __cdecl std::_Xbad_alloc(void)" (?_Xbad_alloc@std@@YAXXZ) referenced in function "char * __cdecl std::_Allocate<char>(unsigned __int64,char *)" (??$_Allocate@D@std@@YAPEAD_KPEAD@Z)
1>inetc.obj : error LNK2019: unresolved external symbol "void __cdecl std::_Xlength_error(char const *)" (?_Xlength_error@std@@YAXPEBD@Z) referenced in function "public: bool __cdecl std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> >::_Grow(unsigned __int64,bool)" (?_Grow@?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@QEAA_N_K_N@Z)
1>inetc.obj : error LNK2019: unresolved external symbol "void __cdecl std::_Xout_of_range(char const *)" (?_Xout_of_range@std@@YAXPEBD@Z) referenced in function "public: void __cdecl std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> >::_Xran(void)const " (?_Xran@?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@QEBAXXZ)
1>inetc.obj : error LNK2019: unresolved external symbol "char const * __cdecl std::_Syserror_map(int)" (?_Syserror_map@std@@YAPEBDH@Z) referenced in function "public: virtual class std::error_condition __cdecl std::_System_error_category::default_error_condition(int)const " (?default_error_condition@_System_error_category@std@@UEBA?AVerror_condition@2@H@Z)
1>inetc.obj : error LNK2019: unresolved external symbol "char const * __cdecl std::_Winerror_map(int)" (?_Winerror_map@std@@YAPEBDH@Z) referenced in function "public: virtual class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > __cdecl std::_System_error_category::message(int)const " (?message@_System_error_category@std@@UEBA?AV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@2@H@Z)
1>inetc.obj : error LNK2019: unresolved external symbol _CxxThrowException referenced in function __catch$?_Copy@?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@QEAAX_K0@Z$1
1>inetc.obj : error LNK2001: unresolved external symbol __CxxFrameHandler3
1>inetc.obj : error LNK2001: unresolved external symbol "const type_info::`vftable'" (??_7type_info@@6B@)
```
I assume that you were able to build it successfully?

---

Reviewed 1 of 1 files at r1.
Review status: all files reviewed at latest revision, all discussions resolved.

---

*[Contrib/Inetc/inetc.cpp, line 1429 at r1](https://reviewable.io/reviews/digitalmediaserver/nsis-inetc-plugin/1#-LB5mDg-8xnOIhLFxQzK:-LB5mDg-8xnOIhLFxQzL:b78dlsy) ([raw file](https://github.com/digitalmediaserver/nsis-inetc-plugin/blob/67654ef7b85be328115b8148e1f54d6be6d5f8fa/Contrib/Inetc/inetc.cpp#L1429)):*
> ```cpp
> 		{
> 			unsigned int fgR, fgG, fgB;
> 			sscanf_s(szTextColor, "%02x%02x%02x", &fgR, &fgG, &fgB);
> ```

This fails to compile with Unicode, ```argument of type TCHAR * is incompatible with parameter of type const char *```

This is the same for all calls to ```sscanf_s()```. I'm not anywhere close to knowing C++ types well enough to find a solution, but hopefully you are.

---


*Comments from [Reviewable](https://reviewable.io/reviews/digitalmediaserver/nsis-inetc-plugin/1#-:-LB5mwJJ5te3ZHWyX3co:b-a44h43)*

**Comment by Nutzzz** [comment]
Re: it being an addition, well, that was the "(For what it's worth)"; seeing as I liked *your* addition, and there was nowhere else to really contribute [though it could be added to the Wiki]... here it is.
Re: not compiling, Release (but not Debug) has "Linker > Input > Ignore All Default Libraries" turned on, but it compiles if you set it back to the default (No).
As for Unicode support, we don't need at the moment for our purposes, so it wasn't on my radar, but it should be an easy fix.  We'll take a look.

**Comment by Nadahar** [comment]
@Nutzzz It seems a word was missing in my response, which made it a bit hard to understand. What I meant to say was 
> his sounds like a *nice* addition

Adding things to this repo is fine by me, I just wanted to make sure that you knew that it wasn't the "official" repo.

As I'm sure you understand, C++ is not my language, and VS is completely alien to me. The default libraries setting makes sense, but it makes me worry about the consequences of referencing one of these "default" libraries. Does that mean that it won't run without VCRedist or (god forbid) .NET installed? If so, I think that is a big limitation because of such a simple task. It that's the case, wouldn't it be better to make a parser for the color values that manages without extra libraries?

**Comment by Nutzzz** [comment]
OK.  Up'd a couple of quick ifdefs for Unicode, and the new project file should allow you to build everything, including x64.

The required libraries are statically linked, so no redistributables are required.

**Comment by Nutzzz** [comment]
FYI, I'm now getting a crash.  Debugging...

File: Contrib/Inetc/inetc.cpp [hunk 1]
```diff
@@ -1456,6 +1458,7 @@ INT_PTR CALLBACK dlgProc(HWND hDlg, UINT message, WPARAM wParam, LPARAM lParam )
     }
 	case WM_CTLCOLORDLG:
 	{
+		szBgColor[6] = 0;
```
Comment (Nadahar) [inline Contrib/Inetc/inetc.cpp]: @Nutzzz As stated before, I'm really not up to speed with C++, but how you can set index 6 of ```szBgColor``` when it according to the next ```if``` might be ```NULL```? Won't that lead to an error if it is in fact ```NULL```?

File: Contrib/Inetc/inetc.cpp [hunk 2]
```diff
@@ -1682,7 +1689,6 @@ void __declspec(dllexport) __cdecl get(HWND hwndParent,
 	pushstring(url);
 //	if(*szCaption == 0) lstrcpy(szCaption, PLUGIN_NAME);
 	if(*szUserAgent == 0) lstrcpy(szUserAgent, INETC_USERAGENT);
-	if(*szPassword && *szUsername)
```
Comment (Nadahar) [inline Contrib/Inetc/inetc.cpp]: @Nutzzz Why did you remove this line?

Comment (Nutzzz) [reply Contrib/Inetc/inetc.cpp]: That was a mistake.  Thanks for catching it.

**Comment by Nutzzz** [comment]
Review status: 0 of 2 files reviewed at latest revision, 2 unresolved discussions.

---

*[Contrib/Inetc/inetc.cpp, line 1429 at r1](https://reviewable.io/reviews/digitalmediaserver/nsis-inetc-plugin/1#-LB5mDg-8xnOIhLFxQzK:-LBRuG3t86NZyKqkUvcT:b-896fix) ([raw file](https://github.com/digitalmediaserver/nsis-inetc-plugin/blob/67654ef7b85be328115b8148e1f54d6be6d5f8fa/Contrib/Inetc/inetc.cpp#L1429)):*
<details><summary><i>Previously, Nadahar (Nadar) wrote…</i></summary><blockquote>

This fails to compile with Unicode, ```argument of type TCHAR * is incompatible with parameter of type const char *```

This is the same for all calls to ```sscanf_s()```. I'm not anywhere close to knowing C++ types well enough to find a solution, but hopefully you are.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/digitalmediaserver/nsis-inetc-plugin/1)*

**Comment by Nutzzz** [comment]
Review status: 0 of 2 files reviewed at latest revision, 2 unresolved discussions.

---

*[Contrib/Inetc/inetc.cpp, line 1461 at r3](https://reviewable.io/reviews/digitalmediaserver/nsis-inetc-plugin/1#-LAzOLVYdEOfbTVuxhU5-r3-1461:-LBRuL256OxlyPMZgDC8:b-896fix) ([raw file](https://github.com/digitalmediaserver/nsis-inetc-plugin/blob/314f71644b2ff47d32960640a30bb3b766e957bb/Contrib/Inetc/inetc.cpp#L1461)):*
<details><summary><i>Previously, Nadahar (Nadar) wrote…</i></summary><blockquote>

@Nutzzz As stated before, I'm really not up to speed with C++, but how you can set index 6 of ```szBgColor``` when it according to the next ```if``` might be ```NULL```? Won't that lead to an error if it is in fact ```NULL```?
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/digitalmediaserver/nsis-inetc-plugin/1)*

**Comment by Nadahar** [comment]
@Nutzzz I've found some problems with your code, but I haven't been able to solve it yet. Since I don't use C++ and isn't used to VS, this is very time consuming. The fact that I have to compile the plugin and then compile NSIS for every test makes it even slower, and I don't have the slightest idea of how to get a debugger (break points, variable inspection) up and running for this I'm pretty much blind as well.

What I have found this far is:
* I'm not convinced that your string initialization ```{ 0 }``` is safe. As far as I can understand, that is only correct in ANSI mode while Unicode requires two "null" values. I'm not sure though, but I think ```TEXT("")``` handles this properly.
* The code doesn't work properly when not setting any of your parameters. You seem to have hardcoded the "default" values, but that doesn't cut it. The default theme colors must be preserved when custom colors aren't specified.
* The code breaks compatibility with older Windows versions. [WM_CTLCOLORSTATIC](https://msdn.microsoft.com/en-us/library/windows/desktop/bb787524(v=vs.85).aspx) is only defined for Vista and later. Since it's not defined for earlier versions, I guess the message will never come. In addition I'm afraid that the constant itself isn't defined and thus might be invalid. Since I know so little about how the compiler works, I don't know if this is simply replaced by the value (0x0138 ?) when compiled or not.
* I'm a bit confused why you set the background color two different places, and why you set the background mode to "TRANSPARENT".
* The MSDN documentation clearly states that the brushes creates should be reused, and must be freed after use. This code seems to create a new brush for every message (and never released), which means it's a memory leak and quite inefficient.

**Comment by Nadahar** [comment]
@Nutzzz After a lot of struggeling with things I do not know, I came up with this solution that I think deals with all my issues: 4b0d480e1cacc90f509340bda2384bb68d7d1efe

I've tested it to the best of my abilites, and it now avoids the memory leak, works on XP and keeps the default colors if no colors are specified. Please not that it's not possible to change only the text color, nothing happens unless background color is also specified.

I'd appreciate if you'd review and test it and give some feedback.

**Comment by Nutzzz** [comment]
My abilities are less than yours, though I have had a bit of help from a more expert coworker.  I hope you've viewed this as a learning experience rather than just a pain in the butt.  Your changes look good to both of us, should cover a much wider audience, and resolve the crashing I had observed with my latest pull requests.

**Comment by Nadahar** [comment]
@Nutzzz Thank you for your feedback. I do know programming, just not in C/C++ :wink: Of course this isn't a "pain in the butt", we both try to learn and share as best we can. I don't think anyone can ask for anything more from volunteers.

It's good to hear that this works for you and that you don't see any problems with the code :+1:

**Comment by Nadahar** [comment]
I know I've been slow to work this out, but as my C/C++ skills is appalling and I've only worked on it a little now and then, it's taken time. I'm finally ready now though, so I'll merge this and then add my fixes in another commit.

**Review by Nadahar — APPROVED** [review]
> Reviewed 1 of 2 files at r2, 1 of 1 files at r5.
> *[Reviewable](https://reviewable.io/reviews/digitalmediaserver/nsis-inetc-plugin/1)* status: :shipit: complete! all files reviewed, all discussions resolved
